### PR TITLE
generic check.py

### PR DESCRIPTION
**Pull request by crmulliner**
- add check.py (a generic check script)
- add android/unpack.sh (unpack android ota for use with check.py)

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -80,9 +80,11 @@ Example for using custom scripts stored in the `scripts` directory:
 PATH=$PATH:./scripts fwanalyzer -cfg system_fwa.toml -in system.img -out system_check_output.json
 ```
 
-The [devices](devices/) folder contains helper scripts for unpacking and dealing with specific devices types and firmware package formats such as [Android](devices/android).
+The [devices/](devices/) folder contains helper scripts for unpacking and dealing with specific devices types and firmware package formats such as [Android](devices/android).
```
Comment (Grazfather) [inline Readme.md]: old typo: 'devices types' -> 'device types'

File: Readme.md [hunk 2]
```diff
@@ -80,9 +80,11 @@ Example for using custom scripts stored in the `scripts` directory:
...
 It also includes general configuration files that can be included in target specific FwAnalyzer configurations.
 
+Check.py in the [devices/](devices) folder provides a universal script to effectively use FwAnalyzer, see [devices/Readme.md](devices/Readme.md) for details. For most people that is what you will use FwAnalyzer with.
```
Comment (Grazfather) [inline Readme.md]: Can we be more descriptive of how it's used? Maybe "This is how most people will invoke FwAnalyzer" or something.

At first glance I don't know what check.py is for.

File: Readme.md [hunk 3]
```diff
@@ -80,9 +80,11 @@ Example for using custom scripts stored in the `scripts` directory:
...
+
 The [scripts](scripts/) folder contains helper scripts that can be called from FwAnalyzer for file content analysis and data extraction.
```
Comment (Grazfather) [inline Readme.md]: If we are being consistent, you want a `/` after the `scripts` url

File: devices/Readme.md [hunk 4]
```diff
@@ -1,3 +1,68 @@
 # Devices
 
 This directory contains support tools and popular checks that can be included in FwAnalyzer configs for multiple targets.
+
+- [Android](android)
+- [generic Linux](generic)
+
+## Check.py
+
+is a universal script to run FwAnalyzer. check.py will unpack (with the help of a unpacker; see below) firmware
```
Comment (Grazfather) [inline devices/Readme.md]: "_check.py_ is a universal script to run FwAnalyzer. It will..."

File: devices/Readme.md [hunk 5]
```diff
@@ -1,3 +1,68 @@
...
+and run fwanalyzer against each of the target filesystems, it will combine all of the reports
+into one big report. In addition it will do some post processing of the filetree files (if present) and
+append the result to the report.
+
+Using check.py is straight forward (the example below is for an Android OTA firmware - make sure you have the required Android unpacking tools installed and added to your PATH, see: [Android](android/Readme.md)):
```
Comment (Grazfather) [inline devices/Readme.md]: I'd put everything in these parens into a paragraph after the example.

File: devices/Readme.md [hunk 6]
```diff
@@ -1,3 +1,68 @@
...
+  --report REPORT       report file
+  --keep-unpacked       keep unpacked data
+  --fwanalyzer-bin FWANALYZER_BIN
+                        path to fwanalyzer binary
+  --fwanalyzer-options FWANALYZER_OPTIONS
+                        options passed to fwanalyzer
+```
+
+The _--keep-unpacked_ option will NOT delete the temp directory that contains the unpacked files.
+Once you have the unpacked directory you can pass it to the _--fw_ option and avoid unpacking the
```
Comment (Grazfather) [inline devices/Readme.md]: "Once you have the unpacked directory you can pass it to the _--fw_ option and avoid"

"Once you have the unpacked directory you can pass it to the _--fw_ option **to** avoid"

File: devices/Readme.md [hunk 7]
```diff
@@ -1,3 +1,68 @@
...
+firmware for each run (e.g. while you test/modify your configuration files). See the example below.
+
+```sh
+check.py --unpacker android/unpack.sh --fw /tmp/tmp987689123 --cfg-path android --cfg-include android --fwanalyzer-bin ../build/fwanalyzer
+```
+
+### unpacker
+
+The unpacker is used by check.py to _unpack_ firmware.
+The unpacker needs to be an executable file, that takes two parameters first the `file` to unpack
```
Comment (Grazfather) [inline devices/Readme.md]: Without an extra blank line this will be joined with the above line.

File: devices/android/unpack.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#!/bin/sh
+
+# -- unpack android OTA --
+
+if [ -z "$1" ]; then
+    echo "syntax: %0 <android_ota.zip>"
```
Comment (Grazfather) [inline devices/android/unpack.sh]: `$0`?

File: devices/check.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import tempfile
+import os
+import os.path
```
Comment (Grazfather) [inline devices/check.py]: shouldn't be needed

File: devices/check.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+import sys
+import argparse
```
Comment (Grazfather) [inline devices/check.py]: let's sort

File: devices/check.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+
+class CheckFirmware:
+    def __init__(self, fwanalyzer="fwanalyzer"):
+        self._tmpdir = tempfile.mktemp()
+        self._unpackdir = os.path.join(self._tmpdir, "unpacked")
+        self._fwanalyzer = fwanalyzer
+        self._unpacked = False
+
+
```
Comment (Grazfather) [inline devices/check.py]: you don't need 2 blank lines for pep8 in a class def

File: devices/check.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        return subprocess.check_call(cmd, shell=True)
+
+
+    def unpack(self, fwfile, unpacker, cfgpath):
+        try:
+            if os.path.exists(os.path.join(fwfile, "unpacked")) and os.path.exists(os.path.join(fwfile, "targets.json")):
+                self._tmpdir = fwfile
+                self._unpackdir = os.path.join(self._tmpdir, "unpacked")
+                print("{0}: is a directory containing an 'unpacked' path, skipping".format(fwfile))
+                cmd = "cat " + os.path.join(fwfile, "targets.json")
```
Comment (Grazfather) [inline devices/check.py]: use `.format`

File: devices/check.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+                self._unpacked = True
+            else:
+                os.mkdir(self._tmpdir)
+                os.mkdir(self._unpackdir)
+                cmd = unpacker + " " + fwfile + " " + cfgpath
```
Comment (Grazfather) [inline devices/check.py]: `.format`

File: devices/check.py [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+            res = subprocess.check_output(cmd, shell=True, cwd=self._unpackdir)
+            targets = json.loads(res.decode('utf-8'))
+            fp = open(os.path.join(self._tmpdir, "targets.json"), "w+")
```
Comment (Grazfather) [inline devices/check.py]: `with open...`

File: devices/check.py [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+            return targets
+        except Exception as e:
+            print("Exception: {0}".format(e))
+            print("can't load targets from output of '{0}' check your script".format(unpacker))
+            return None
+
+
+    def del_tmp_dir(self):
+        if not self._unpacked:
+            cmd = "rm -rf " + self._tmpdir
```
Comment (Grazfather) [inline devices/check.py]: `.format`

File: devices/check.py [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+            return subprocess.check_call(cmd, shell=True)
+
+
+    def files_by_ext_stat(self, data):
+        allext = {}
+        for i in data["files"]:
+            fn, ext = os.path.splitext(i["name"])
+            if ext in allext:
+                (count, ext) = allext[ext]
```
Comment (Grazfather) [inline devices/check.py]: no parens needed

File: devices/check.py [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+                allext[ext] = (count + 1, ext)
+            else:
+                allext[ext] = (1, ext)
+        return (len(data["files"]), allext)
```
Comment (Grazfather) [inline devices/check.py]: no parens needed

File: devices/check.py [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+
+    def analyze_filetree(self, filetreefile):
+        fp = open(filetreefile)
```
Comment (Grazfather) [inline devices/check.py]: `with open...`

File: devices/check.py [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+        num_files, stats = self.files_by_ext_stat(data)
+        out = {}
+        percent = num_files / 100
+        # only keep entries with count > 1% and files that have an extension
+        for i in stats:
+            (count, ext) = stats[i]
+            if count > percent and ext != "":
+                out[ext] = (count, ext)
+
+        ftresult = {}
```
Comment (Grazfather) [inline devices/check.py]:
```python
return {
    "total_files": num_files,
    ...
}
```

File: devices/check.py [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+    report["firmware"] = fwfile
+    if os.path.isfile(fwfile):
+        report["firmware_digest"] = hashfile(fwfile)
+    report["status"] = status
+    return json.dumps(report, sort_keys=True, indent=2)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--fw', action='store', required=True, help="path to firmware file OR path to unpacked firmware")
```
Comment (Grazfather) [inline devices/check.py]: let's consistently use `"` instead of `'`

File: devices/check.py [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+    parser.add_argument('--fwanalyzer-options', action='store', default="", help="options passed to fwanalyzer")
+    args = parser.parse_args()
+
+    fw = os.path.realpath(args.fw)
+    cfg = os.path.realpath(args.cfg_path)
+
+    check = CheckFirmware(args.fwanalyzer_bin)
+    print("using tmp directory: {0}".format(check.get_tmp_dir()))
+    targets = check.unpack(fw, os.path.realpath(args.unpacker), cfg)
+    if targets == None or len(targets) == 0:
```
Comment (Grazfather) [inline devices/check.py]: I think `if not targets` covers this

File: devices/check.py [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+            all_checks_ok = False
+
+    if args.keep_unpacked:
+        print("unpacked: {0}\n".format(check.get_tmp_dir()))
+    else:
+        check.del_tmp_dir()
+
+    report = make_report(args.fw, out)
+    if args.report != None:
+        fp = open(args.report, "w+")
```
Comment (Grazfather) [inline devices/check.py]: `with open...`

File: devices/android/unpack.sh (hunk 8 above)
Comment (crmulliner) [reply devices/android/unpack.sh]: the name of the script itself is $0

Comment (Grazfather) [reply devices/android/unpack.sh]: You used `%`

File: devices/check.py [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+    if args.keep_unpacked:
+        print("unpacked: {0}\n".format(check.get_tmp_dir()))
+    else:
+        check.del_tmp_dir()
+
+    report = make_report(args.fw, out)
+    if args.report != None:
+        with open(args.report, "w+") as fp:
+            fp.write(report)
+            fp.close()
```
Comment (Grazfather) [inline devices/check.py]: No needed when you use `with`

File: devices/check.py [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+                self._unpacked = True
+            else:
+                os.mkdir(self._tmpdir)
+                os.mkdir(self._unpackdir)
+                cmd = "{0} {1} {2}".format(unpacker, fwfile, cfgpath)
+            res = subprocess.check_output(cmd, shell=True, cwd=self._unpackdir)
+            targets = json.loads(res.decode('utf-8'))
+            with open(os.path.join(self._tmpdir, "targets.json"), "w+") as fp:
+                fp.write(res.decode('utf-8'))
+                fp.close()
```
Comment (Grazfather) [inline devices/check.py]: Not needed

File: devices/check.py [hunk 25]
```diff
@@ -0,0 +1,205 @@
...
+            else:
+                allext[ext] = (1, ext)
+        return (len(data["files"]), allext)
+
+    def analyze_filetree(self, filetreefile):
+        with open(filetreefile) as fp:
+            data = json.load(fp)
+        num_files, stats = self.files_by_ext_stat(data)
+        out = {}
+        percent = num_files / 100
```
Comment (jlarimer) [inline devices/check.py]: does this work? if num_files is an int then percent will end up being 0 if num_files is < 100

File: devices/check.py [hunk 26]
```diff
@@ -0,0 +1,205 @@
...
+        if "current_file_tree_path" in data:
+            if os.path.isfile(data["current_file_tree_path"]):
+                data["file_tree_analysis"] = self.analyze_filetree(data["current_file_tree_path"])
+
+        return (status, json.dumps(data, sort_keys=True, indent=2))
+
+
+def hashfile(fpath):
+    m = hashlib.sha256()
+    with open(fpath, 'rb') as f:
```
Comment (jlarimer) [inline devices/check.py]: use double quotes for "rb" to be consistent

File: devices/check.py [hunk 27]
```diff
@@ -0,0 +1,205 @@
...
+        with open(result) as read_file:
+            data = json.load(read_file)
+
+        if "offenders" in data:
+            status = False
+        else:
+            status = True
+
+        if "current_file_tree_path" in data:
+            if os.path.isfile(data["current_file_tree_path"]):
```
Comment (jlarimer) [inline devices/check.py]: const for "current_file_tree_path"

File: devices/check.py (hunk 25 above)
Comment (crmulliner) [reply devices/check.py]: yes and then we get stats for every file extension :)

File: devices/check.py [hunk 28]
```diff
@@ -0,0 +1,210 @@
...
+        with open(args.report, "w+") as fp:
+            fp.write(report)
+        print("report written to '{0}'".format(args.report))
+    else:
+        print(report)
+
+    if not all_checks_ok:
+        print("Firmware Analysis: checks failed")
+        sys.exit(1)
+    else:
```
Comment (Grazfather) [inline devices/check.py]: no need for this else technically

**Pull request merged**